### PR TITLE
Create AudioStreamPlayer when dropping AudioStream

### LINKED ISSUE #9699: Create an `AudioStreamPlayer` when an `AudioStream` is dropped into the scene
### Describe the project you are working on

Quick game prototypes

### Describe the problem or limitation you are having in your project

Adding sounds to the game could be a bit quicker

In Unity dropping a sound file into the hierarchy automatically creates a player for it. This QOL feature is convenient for quick game prototypes where you're not using any kind of a centralized audio manager singleton to play streams through, or if you're purposefully avoiding having such a singleton if you're trying to make game entities more self-contained

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

When you drop an `AudioStream` resource from the FileSystem dock onto a node in the Scene dock or into the viewport it would automatically create an appropriate `AudioStreamPlayer` node and set its `stream` property to that resource

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The type of the node would depend on where you're dropping the stream:

- Dropping onto a node derived from `Node3D` or into the 3D viewport would create an `AudioStreamPlayer3D`
- Dropping onto a node derived from `Node2D` or into the 2D viewport would create an `AudioStreamPlayer2D`
- Dropping anywhere else would create an `AudioStreamPlayer`

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not in a few

### Is there a reason why this should be core and not an add-on in the asset library?

A small QOL editor feature

### PR DESCRIPTION
**Pull request by timothyqiu**
Closes https://github.com/godotengine/godot-proposals/issues/9699

- Create AudioStreamPlayer if dropped in between nodes in the Scene dock
- Create AudioStreamPlayer2D if dropped into 2D editor
- Create AudioStreamPlayer3D if dropped into 3D editor

https://github.com/godotengine/godot/assets/372476/8bb2a5fe-d8bb-4b28-9273-a32d775733ec

### CODE REVIEW

**Comment by timothyqiu** [comment]
> Is there a reason not to check if the parent node is derived from Node2D/Node3D when dropping into the scene dock to instantiate an appropriate player like I mentioned in the proposal?

Oh, using the parent node as a basis is a good idea.

https://github.com/godotengine/godot-proposals/issues/9699 is about dropping _onto_ a node. That action is taken by property setter.

> Would it be hard to cast a ray when dragging it into the 3d editor to be able to place it on colliders?

It already is.

> I can also think of some potential cases where you won't have the viewport available at all, only docks

AFAIK, the debugger does not support assigning resource properties at runtime. i.e., you can create the audio stream player node, but its stream property won't be synced. That's a current limitation.

**Comment by passivestar** [comment]
> There should be a way to toggle the behavior

A modifier key perhaps?

> AFAIK, the debugger does not support assigning resource properties at runtime. i.e., you can create the audio stream player node, but its stream property won't be synced. That's a current limitation

I believe it does support that, but I'll open a proposal in the future if that's not the case. I've been doing some R&D related to hot-reloading at runtime (with a GDScript plugin) and it doesn't seem hard to achieve those things anyway

> It already is.

My bad, I see now that I made a mistake while testing

**Comment by timothyqiu** [comment]
- Made the 3D viewport preview not fixed size, so it's clear how it interacts with the environment.
- For the Scene dock, holding Shift while dropping creates AudioStreamPlayer2D for Node2D parent and creates AudioStreamPlayer3D for Node3D parent.
    - Note that if a node does not have any child, you can't drop the new node as its child. That's a limitation of Tree (it only supports dropping above node / below node / on node, no horizontal detection). When moving nodes inside the tree, dropping node A onto node B makes A the child of B. But for files, this action is already taken.


https://github.com/godotengine/godot/assets/372476/078c644f-ccde-4187-9c81-0debbb329142

**Comment by passivestar** [comment]
> Note that if a node does not have any child, you can't drop the new node as its child.

But I can though, wdym?

https://github.com/godotengine/godot/assets/60579014/405224d8-64ab-469a-90d7-a43774a1d9d7

**Comment by timothyqiu** [comment]
Looks like I accidentally broken the "drop to set property" feature :joy:

**Comment by passivestar** [comment]
> Looks like I accidentally broken the "drop to set property" feature 😂

Can you leave it "broken" when another modifier key is held? It's kinda super convenient :D

**Comment by KoBeWi** [comment]
> Or perhaps do what we discussed in the proposal and check if the node has a "stream" property to assign to?

Valid properties are only checked at the time of drop, to avoid performance problems when hovering.

**Comment by passivestar** [comment]
> Valid properties are only checked at the time of drop, to avoid performance problems when hovering.

Doesn't it work for this PR? There's no preview in the scene dock or anything like that

**Comment by timothyqiu** [comment]
I cleaned up the drop code a bit.

Now scene tree dock always tries to do a property drop first. If no property can be set, then the file is dropped as a new node.

This has a side-effect: Say a node has a `PackedScene` property and you're dropping a scene file onto that node. It instantiates the scene as child before, and now it sets the property.

I think this change is okay because:

- It makes dropping behavior more consistent.
- No special case handling is involved.
- This situation is relatively rare.

If this concerns, maybe we can add an editor setting like `docks/scene_tree/prefer_drop_as_property` to swap the priority.

**Comment by timothyqiu** [comment]
> When dropping a scene right under a node that has a `PackedScene` property it would assign it to the property.

Fixed. Turns out to be my mistake, sorry :P

**Comment by KoBeWi** [comment]
Audio dropping is missing header.
image
vs
image

File: editor/scene_tree_dock.cpp [hunk 1]
```diff
@@ -3238,10 +3304,22 @@ void SceneTreeDock::_files_dropped(const Vector<String> &p_files, NodePath p_to,
 			menu_properties->reset_size();
 			menu_properties->set_position(get_screen_position() + get_local_mouse_position());
 			menu_properties->popup();
-		} else if (!valid_properties.is_empty()) {
+			return;
+		}
+		if (!valid_properties.is_empty()) {
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:3309]: Why did you remove `else if`?

File: editor/scene_tree_dock.cpp [hunk 2]
```diff
@@ -330,6 +331,73 @@ void SceneTreeDock::_perform_instantiate_scenes(const Vector<String> &p_files, N
...
+			current_option = -1;
+			accept->set_text(vformat(TTR("Error loading audio stream from %s"), p_files[i]));
+			accept->popup_centered();
+			error = true;
+			break;
+		}
+
+		Node *player = Object::cast_to<Node>(ClassDB::instantiate(node_type));
+		player->set("stream", stream);
+		player->set_name(Node::adjust_name_casing(p_files[i].get_file().get_basename().validate_node_name()));
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:361]: This can result in empty name, e.g. when file is named `%.mp3`. It's a corner-case, but results in an error.

File: editor/scene_tree_dock.cpp [hunk 3]
```diff
@@ -330,6 +331,73 @@ void SceneTreeDock::_perform_instantiate_scenes(const Vector<String> &p_files, N
...
+
+	for (int i = 0; i < nodes.size(); i++) {
+		Node *node = nodes[i];
+
+		undo_redo->add_do_method(parent, "add_child", node, true);
+		if (p_pos >= 0) {
+			undo_redo->add_do_method(parent, "move_child", node, p_pos + i);
+		}
+		undo_redo->add_do_method(node, "set_owner", edited_scene);
+		undo_redo->add_do_method(editor_selection, "clear");
```
Comment (KoBeWi) [inline editor/scene_tree_dock.cpp:383]: This clears selection every time. You should do it before the loop, so all nodes are selected.

File: editor/plugins/node_3d_editor_plugin.cpp [hunk 4]
```diff
@@ -4195,24 +4197,34 @@ void Node3DEditorViewport::_create_preview_node(const Vector<String> &files) con
...
+		if (scene.is_valid()) {
+			Node *instance = scene->instantiate();
+			if (instance) {
+				instance = _sanitize_preview_node(instance);
+				preview_node->add_child(instance);
 			}
 			add_preview = true;
 		}
+
+		Ref<Mesh> mesh = Ref<Mesh>(Object::cast_to<Mesh>(*res));
```
Comment (KoBeWi) [inline editor/plugins/node_3d_editor_plugin.cpp:4211]:
```suggestion
		Ref<Mesh> mesh = res;
```
This is how you cast Resources. idk why the code was like that.

File: editor/plugins/node_3d_editor_plugin.cpp [hunk 5]
```diff
@@ -4419,6 +4431,47 @@ bool Node3DEditorViewport::_create_instance(Node *parent, String &path, const Po
 	return true;
 }
 
+bool Node3DEditorViewport::_create_audio_node(Node *parent, String &path, const Point2 &p_point) {
+	Ref<Resource> res = ResourceLoader::load(path);
```
Comment (KoBeWi) [inline editor/plugins/node_3d_editor_plugin.cpp:4435]:
```suggestion
	Ref<AudioStream> audio = ResourceLoader::load(path);
```
You can assign it directly here. It will be null if type does not match.

File: editor/plugins/node_3d_editor_plugin.cpp [hunk 6]
```diff
@@ -4419,6 +4431,47 @@ bool Node3DEditorViewport::_create_instance(Node *parent, String &path, const Po
...
+	audio_player->set_name(Node::adjust_name_casing(name));
+
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->add_do_method(parent, "add_child", audio_player, true);
+	undo_redo->add_do_method(audio_player, "set_owner", EditorNode::get_singleton()->get_edited_scene());
+	undo_redo->add_do_reference(audio_player);
+	undo_redo->add_undo_method(parent, "remove_child", audio_player);
+	undo_redo->add_do_method(editor_selection, "add_node", audio_player);
+
+	String new_name = parent->validate_child_name(audio_player);
```
Comment (KoBeWi) [inline editor/plugins/node_3d_editor_plugin.cpp:4455]:
```suggestion
	const String new_name = parent->validate_child_name(audio_player);
```

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 7]
```diff
@@ -5717,46 +5719,57 @@ void CanvasItemEditorViewport::_on_change_type_confirmed() {
...
 		}
+
+		Ref<AudioStream> audio = Ref<AudioStream>(Object::cast_to<AudioStream>(*res));
+		if (audio.is_valid()) {
+			Sprite2D *sprite = memnew(Sprite2D);
+			sprite->set_texture(get_editor_theme_icon(SNAME("AudioStreamPlayer2D")));
+			sprite->set_modulate(Color(1, 1, 1, 0.7f));
+			sprite->set_position(Vector2(0, -sprite->get_texture()->get_size().height) * EDSCALE);
+			preview_node->add_child(sprite);
+			label_desc->set_text(desc);
```
Comment (KoBeWi) [inline editor/plugins/canvas_item_editor_plugin.cpp:5769]: You could mention Shift to add AudioStreamPlayer2D.

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 8]
```diff
@@ -5717,46 +5719,57 @@ void CanvasItemEditorViewport::_on_change_type_confirmed() {
...
+		ERR_CONTINUE(res.is_null());
+
 		Ref<Texture2D> texture = Ref<Texture2D>(Object::cast_to<Texture2D>(*res));
+		if (texture.is_valid()) {
+			Sprite2D *sprite = memnew(Sprite2D);
+			sprite->set_texture(texture);
+			sprite->set_modulate(Color(1, 1, 1, 0.7f));
+			preview_node->add_child(sprite);
+			label->show();
+			label_desc->set_text(desc + "\n" + TTR("Hold Alt + Shift when dropping to add as a different node type."));
```
Comment (KoBeWi) [inline editor/plugins/canvas_item_editor_plugin.cpp:5746]: This is a pre-existing issue, but since this runs in a for loop, the resulting text can be a combination of strings that depends on order of files. E.g. texture header without type mention:
image

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 9]
```diff
@@ -5863,6 +5876,51 @@ void CanvasItemEditorViewport::_create_nodes(Node *parent, Node *child, String &
...
+		EditorDebuggerNode *ed = EditorDebuggerNode::get_singleton();
+		undo_redo->add_do_method(ed, "live_debug_create_node", EditorNode::get_singleton()->get_edited_scene()->get_path_to(parent), child->get_class(), new_name);
+		undo_redo->add_undo_method(ed, "live_debug_remove_node", NodePath(String(EditorNode::get_singleton()->get_edited_scene()->get_path_to(parent)) + "/" + new_name));
+	}
+
+	// Compute the global position
+	Transform2D xform = canvas_item_editor->get_canvas_transform();
+	Point2 target_position = xform.affine_inverse().xform(p_point);
+
+	// there's nothing to be used as source position so snapping will work as absolute if enabled
```
Comment (KoBeWi) [inline editor/plugins/canvas_item_editor_plugin.cpp:5912]:
```suggestion
	// There's nothing to be used as source position, so snapping will work as absolute if enabled.
```

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 10]
```diff
@@ -5973,9 +6031,14 @@ void CanvasItemEditorViewport::_perform_drop_data() {
...
 		if (texture.is_valid()) {
 			Node *child = Object::cast_to<Node>(ClassDB::instantiate(default_texture_node_type));
-			_create_nodes(target_node, child, path, drop_pos);
+			_create_texture_node(target_node, child, path, drop_pos);
 			undo_redo->add_do_method(editor_selection, "add_node", child);
 		}
+
+		Ref<AudioStream> audio = res;
+		if (audio.is_valid()) {
+			_create_audio_node(target_node, path, drop_pos);
```
Comment (KoBeWi) [inline editor/plugins/canvas_item_editor_plugin.cpp:6040]: I don't get why this method takes path instead of the resource it just loaded. (I know it's the same as above, I just wonder)

File: editor/scene_tree_dock.cpp (hunk 1 above)
Comment (timothyqiu) [reply editor/scene_tree_dock.cpp:3309]: These two branches both returns now. So the `else` is redundant.

File: editor/scene_tree_dock.cpp (hunk 2 above)
Comment (timothyqiu) [reply editor/scene_tree_dock.cpp:361]: Dropping texture as Sprite2D has a similar issue for `_.mp3`. I'll try to fix both.

File: editor/scene_tree_dock.cpp (hunk 3 above)
Comment (timothyqiu) [reply editor/scene_tree_dock.cpp:383]: This section is copied from dropping scenes. I'll fix both.

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 7 above)
Comment (timothyqiu) [reply editor/plugins/canvas_item_editor_plugin.cpp:5769]: Dropping audio files into the 2D viewport always creates AudioStreamPlayer2D. Modifier keys are for adjusting the new node's position.

**Comment by timothyqiu** [comment]
> Audio dropping is missing header.

Currently, the header is shown if the file list contains texture. I added the same for audio streams in the recent update. e.g.:

- Adding Sprite2D
- Adding Sprite2D and AudioStreamPlayer2D

Note that scene-dropping does not have this header either. I'm out of ideas about how to describe the combinations :stuck_out_tongue:  Probably worth another PR.

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 10 above)
Comment (timothyqiu) [reply editor/plugins/canvas_item_editor_plugin.cpp:6029]: The name of the node is taken from the path of the resource. Although it is possible to get from `Resource.resource_path`.

Comment (KoBeWi) [reply editor/plugins/canvas_item_editor_plugin.cpp:6029]: Yep.

**Comment by Nodragem** [comment]
> Audio dropping is missing header. 

Hello, I open a proposal on making the dropping tooltips consistent between object types and 2D/3D Viewport a few days ago: https://github.com/godotengine/godot-proposals/issues/9752

I am going to start working on it this weekend and open a PR. So maybe we could add the Audio dropping tooltip to that PR rather than to this one? to avoid merge conflict?

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 11]
```diff
@@ -5922,12 +5973,8 @@ void CanvasItemEditorViewport::_perform_drop_data() {
...
 
 	if (!target_node) {
-		// Without root dropping multiple files is not allowed
-		if (selected_files.size() > 1) {
-			accept->set_text(TTR("Cannot instantiate multiple nodes without root."));
-			accept->popup_centered();
-			return;
-		}
+		// Should already be handled by `can_drop_data`.
+		ERR_FAIL_COND_MSG(selected_files.size() > 1, "Can't instantiate multiple scenes without root.");
```
Comment (KoBeWi) [inline editor/plugins/canvas_item_editor_plugin.cpp:5977]: Why did you change it to say "scenes"? Sprites from texture are not scenes.

**Review by KoBeWi — APPROVED** [review]

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 11 above)
Comment (timothyqiu) [reply editor/plugins/canvas_item_editor_plugin.cpp]: Ah, I used `ERR_FAIL_COND()` here initially. The message was written inside `can_drop_data()`, but also became incorrect after multiple passes of refactoring. Then I finally copied the message here, it's not intentional :joy:

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by patwork** [comment]
@timothyqiu hello, after this commit I'm getting `Error loading audio stream from icon.svg`. when I'm dragging default godot icon.svg into scene tree...

err
